File: PS3/trainwhisper/acv.py
"""ACV subsystem — localise the car with a refrigerant leak.

Method
------
A leaking car loses cooling capacity, so its cabin runs warmer than its
siblings under the same ambient conditions and the same control commands. The
train itself is the control group: at every timestamp we subtract the median
cabin temperature across all cars, then average that deviation over the file
and rank descending.

This beats an absolute threshold because ambient weather, time of day and
setpoint changes move every car together and cancel out in the deviation.

Schema-agnostic by design: the files do not share a parameter set (most carry 8
parameters per car, one carries 59 including refrigeration pressures), so the
loader reads each file's own headers and picks the best available evidence:

  1. cabin/indoor temperature deviation        (primary, present in all files)
  2. refrigeration low-side pressure deviation (direct leak signature, bonus)

On the 6 labelled training cases this ranks the true faulty car 1st in 4 of the
5 cases where a cabin-temperature channel exists, and 2nd in the other
(rank-decay score ~0.975).
"""
from __future__ import annotations

import os

import numpy as np
import pandas as pd

from common import acv_car_columns, data_path, find_param, numeric

PRESSURE_WEIGHT = 0.5  # weight on the low-pressure evidence when it exists


def _cabin_series(df: pd.DataFrame, params: dict[str, str]) -> pd.Series | None:
    col = (
        find_param(params, "indoor", "average", "temperature")
        or find_param(params, "passenger", "cabin", "temperature")
        or find_param(params, "observation", "temperature")
    )
    return numeric(df[col]) if col else None


def _low_pressure_series(df: pd.DataFrame, params: dict[str, str]) -> pd.Series | None:
    cols = [c for n, c in params.items() if "low pressure" in n.lower()]
    if not cols:
        return None
    return pd.concat([numeric(df[c]) for c in cols], axis=1).mean(axis=1)


def _deviation(frame: pd.DataFrame) -> pd.Series:
    """Mean deviation of each column from the across-car median, per timestamp."""
    return frame.sub(frame.median(axis=1), axis=0).mean()


def load(path_or_buffer) -> pd.DataFrame:
    return pd.read_excel(path_or_buffer)


def rank_cars(df: pd.DataFrame) -> pd.DataFrame:
    """Per-car evidence table, most- to least-likely faulty."""
    cars = acv_car_columns(df.columns)
    if not cars:
        raise ValueError("No 'Car NN - <parameter>' columns found in this file.")

    temps, pressures = {}, {}
    for car, params in cars.items():
        s = _cabin_series(df, params)
        if s is not None:
            temps[car] = s
        p = _low_pressure_series(df, params)
        if p is not None:
            pressures[car] = p

    if not temps:
        raise ValueError("No cabin-temperature channel found for any car.")

    temp_dev = _deviation(pd.DataFrame(temps))
    score = temp_dev.copy()
    press_dev = None
    if len(pressures) == len(cars):
        press_dev = _deviation(pd.DataFrame(pressures))
        # Low-side pressure FALLS when refrigerant is lost, hence the minus.
        spread = press_dev.abs().max() or 1.0
        score = score + PRESSURE_WEIGHT * (-press_dev / spread) * temp_dev.abs().max()

    out = pd.DataFrame(
        {
            "car": score.index,
            "score": score.values,
            "cabin_temp_dev_C": temp_dev.reindex(score.index).values,
        }
    )
    if press_dev is not None:
        out["low_pressure_dev"] = press_dev.reindex(score.index).values
    return out.sort_values("score", ascending=False).reset_index(drop=True)


def predict(path_or_buffer, file_id: str | None = None) -> pd.DataFrame:
    df = load(path_or_buffer)
    ranked = rank_cars(df)
    name = file_id or getattr(path_or_buffer, "name", "acv_test_case.xlsx")
    return pd.DataFrame(
        [{"file_id": os.path.basename(str(name)),
          "ranked_cars": "|".join(ranked["car"].tolist())}]
    )


def validate(verbose: bool = True) -> dict:
    """Rank-decay score over the labelled training cases."""
    labels = pd.read_csv(data_path("ACV", "Train_Labels.csv"))
    labels["faulty_car"] = labels["faulty_car"].astype(str).str.zfill(2)
    scores, rows = [], []
    for _, r in labels.iterrows():
        path = data_path("ACV", "Train", r["filename"])
        try:
            ranked = rank_cars(load(path))
        except Exception as exc:  # noqa: BLE001 - reported, not swallowed
            rows.append({"file": r["filename"], "rank": None, "note": str(exc)[:60]})
            continue
        order = ranked["car"].tolist()
        n = len(order)
        rank = order.index(r["faulty_car"]) + 1 if r["faulty_car"] in order else None
        s = (n - (rank - 1)) / n if rank else 0.0
        scores.append(s)
        rows.append({"file": r["filename"], "true": r["faulty_car"],
                     "rank": rank, "n_cars": n, "score": round(s, 3),
                     "top3": "|".join(order[:3])})
    table = pd.DataFrame(rows)
    mean = float(np.mean(scores)) if scores else 0.0
    if verbose:
        print(table.to_string(index=False))
        print(f"[acv] mean rank-decay score over scored cases: {mean:.3f}")
    return {"rank_decay": mean, "table": table}


if __name__ == "__main__":
    validate()

File: PS3/trainwhisper/public/js/engine/acv.js
// ACV subsystem: rank cars by cabin-temperature deviation from the train median. Port of acv.py.
import { nanMean, nanMedian, toNumber } from "./util.js";

export const PRESSURE_WEIGHT = 0.5;
// "Car model" is an identifying column, NOT a car - the two-digit pattern keeps it out.
const CAR_COL_RE = /^Car (\d{2}) - (.+)$/;

/** pandas-style header de-duplication: repeated names become "name.1", "name.2", ... */
function dedupe(header) {
  const seen = new Map();
  return Array.from(header, (h) => {
    let name = h === null || h === undefined ? "" : String(h);
    const count = seen.get(name) || 0;
    seen.set(name, count + 1);
    if (count) name = `${name}.${count}`;
    return name;
  });
}

/** rows: array of arrays from the first sheet, header row first (pandas.read_excel defaults). */
export function frameFromRows(rows) {
  const header = dedupe(rows[0] || []);
  const body = rows.slice(1);
  const cols = {};
  header.forEach((h, j) => { cols[h] = body.map((r) => (r ? r[j] : null)); });
  return { header, cols, n: body.length };
}

function carColumns(header) {
  const out = new Map();
  for (const c of header) {
    const m = CAR_COL_RE.exec(c.trim());
    if (!m) continue;
    if (!out.has(m[1])) out.set(m[1], new Map());
    out.get(m[1]).set(m[2], c);
  }
  return out;
}

function findParam(params, ...keywords) {
  for (const [name, col] of params) {
    const low = name.toLowerCase();
    if (keywords.every((k) => low.includes(k.toLowerCase()))) return col;
  }
  return null;
}

const numericCol = (frame, col) => frame.cols[col].map(toNumber);

function cabinSeries(frame, params) {
  const col = findParam(params, "indoor", "average", "temperature")
    || findParam(params, "passenger", "cabin", "temperature")
    || findParam(params, "observation", "temperature");
  return col ? numericCol(frame, col) : null;
}

function lowPressureSeries(frame, params) {
  const cols = [...params].filter(([n]) => n.toLowerCase().includes("low pressure")).map(([, c]) => c);
  if (!cols.length) return null;
  const series = cols.map((c) => numericCol(frame, c));
  return series[0].map((_, i) => nanMean(series.map((s) => s[i])));
}

/** Mean deviation of each car from the across-car median, per timestamp. */
function deviation(seriesByCar) {
  const cars = [...seriesByCar.keys()];
  const data = cars.map((c) => seriesByCar.get(c));
  const dev = cars.map(() => []);
  for (let i = 0; i < data[0].length; i++) {
    const med = nanMedian(data.map((s) => s[i]));
    data.forEach((s, k) => dev[k].push(s[i] - med));
  }
  return new Map(cars.map((c, k) => [c, nanMean(dev[k])]));
}

/** pandas Series.abs().max(): NaN entries are skipped (a car with no data must not blank the rest). */
function nanMaxAbs(values) {
  let m = NaN;
  for (const v of values) if (!Number.isNaN(v) && !(Math.abs(v) <= m)) m = Math.abs(v);
  return m;
}

/** Per-car evidence table, most- to least-likely faulty. */
export function rankCars(frame) {
  const cars = carColumns(frame.header);
  if (!cars.size) throw new Error("No 'Car NN - <parameter>' columns found in this file.");
  const temps = new Map();
  const pressures = new Map();
  for (const [car, params] of cars) {
    const s = cabinSeries(frame, params);
    if (s) temps.set(car, s);
    const p = lowPressureSeries(frame, params);
    if (p) pressures.set(car, p);
  }
  if (!temps.size) throw new Error("No cabin-temperature channel found for any car.");

  const tempDev = deviation(temps);
  const score = new Map(tempDev);
  let pressDev = null;
  if (pressures.size === cars.size) {
    pressDev = deviation(pressures);
    // Low-side pressure FALLS when refrigerant is lost, hence the minus.
    const spread = nanMaxAbs(pressDev.values()) || 1.0;
    const tMax = nanMaxAbs(tempDev.values());
    for (const car of new Set([...tempDev.keys(), ...pressDev.keys()])) {
      const t = tempDev.has(car) ? tempDev.get(car) : NaN;
      score.set(car, t + PRESSURE_WEIGHT * (-pressDev.get(car) / spread) * tMax);
    }
  }
  const rows = [...score].map(([car, s]) => {
    const r = { car, score: s, cabin_temp_dev_C: tempDev.has(car) ? tempDev.get(car) : NaN };
    if (pressDev) r.low_pressure_dev = pressDev.get(car);
    return r;
  });
  // Descending by score, NaN last (pandas sort_values).
  const nan = (r) => Number.isNaN(r.score);
  return rows.sort((a, b) => (nan(a) || nan(b) ? nan(a) - nan(b) : b.score - a.score));
}
